src/main/java/com/code_intelligence/jazzer/utils/ZipUtils.java

```diff
@@ -0,0 +1,110 @@
+ * limitations under the License.
+ */
+
+package com.code_intelligence.jazzer.utils;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.IllegalArgumentException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+public class ZipUtils {
+  public static HashSet<String> mergeZipToZip(String src, ZipOutputStream zos,
+      HashSet<String> skipFiles) throws FileNotFoundException, IOException {
+    JarFile jarFile = new JarFile(src);
+
+    // Copy entries from src to dst (jarFile to ZipOutputStream)
+    Enumeration allEntries = jarFile.entries();
+    HashSet<String> filesAdded = new HashSet<String>();
+    while (allEntries.hasMoreElements()) {
+      InputStream is = null;
+      try {
+        ZipEntry entry = (ZipEntry) allEntries.nextElement();
+        if (skipFiles != null && skipFiles.contains(entry.getName())) {
+          continue;
+        }
+
+        zos.putNextEntry(new ZipEntry(entry.getName()));
+        is = jarFile.getInputStream(entry);
+        byte[] buf = new byte[1024];
+        int i = 0;
+        while ((i = is.read(buf)) != -1) {
+          zos.write(buf, 0, i);
+        }
+        zos.closeEntry();
+        filesAdded.add(entry.getName());
+      } finally {
+        if (is != null) {
+          is.close();
+        }
+      }
+    }
+
+    return filesAdded;
+  }
+
+  public static HashSet<String> mergeInstrumentedDirectoryToZip(
+      String src, ZipOutputStream zos, HashSet<String> skipFiles)
+      throws IllegalArgumentException, FileNotFoundException, IOException {
+    HashSet<String> filesAdded = new HashSet<String>();
+    File sourceDir = new File(src);
+    if (!sourceDir.isDirectory()) {
+      throw new IllegalArgumentException("Argument src must be a directory.");
+    }
+
+    Files.walkFileTree(sourceDir.toPath(), new SimpleFileVisitor<Path>() {
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        if (file.toString().endsWith(".original.class")) {
+          // Don't copy original, uninstrumented file
+          return FileVisitResult.CONTINUE;
+        }
+        if (file.toString().endsWith(".failed.class")) {
+          // Don't copy failed file
+          return FileVisitResult.CONTINUE;
+        }
+
+        String zipPath = sourceDir.toPath().relativize(file).toString();
+        if (skipFiles != null && skipFiles.contains(zipPath)) {
```

Comment:
This would block some files that we didn't want to block, ex: Skip files has "Parse.java" then we also have "XmlParse.java" we would skip adding both, when we only want to skip adding "Parse.java"

Reply:
We probably misunderstood each other. My suggestion was to add `".original.class` and `".failed.class` to the `skipFiles` parameter at the invocation of `mergeDirectoryToZip`in `OfflineInstrumentor::createInstrumentedJar`. That would remove all Jazzer instrumentation specific code from `mergeDirectoryToZip`, so that it would better fit it utility characteristic.
This should also not interfere with other blocking cases, like the one you mentioned, right?

Reply:
I see, yes you are 100% right. Updated to your suggestion